### ISSUE TITLE
In what case does `SSL_accept()` return (-1, WANT_READ) without processing pending TLS record?

### ISSUE THREAD

**Issue by cwyang** [openssl/openssl#12336]
Hi,

When `SSL_accept()` returns -1 with WANT_READ, I expect openssl process all pending TLS records which reside in BIO buffer at the time of calling `SSL_accept()`. However I find my expectation wrong in some cases.

Infinite loop happens with the following code.
* Normally the input buffer is completely consumed after `SSL_accept()`. (`buffer size: 0` is printed)
* Sometimes infinite loop happens, and TLS alert record remains in the buffer without processing. (`buffer size: 7` is printed and payload is as follows)
```
00000000 15 03 01 00 02 02 30                            ......0
...
00000000 15 03 03 00 02 02 30                            ......0
```
* It does not mean that every TLS alert from clients triggers infinite loop. Normally TLS alerts are processed and that 7 bytes are consumed. (`buffer size: 0` is printed)

In what case does `SSL_accept()` return (-1, WANT_READ) without processing pending TLS record?  Am I missing something important?

```
retry:
rc = SSL_accept(ssl);
ssl_rc = SSL_get_error(ssl, rc);
...
if (rc < 0 && ssl_rc == SSL_ERROR_WANT_READ) {
  // inputbuf is linked to BIO and consumed when BIO_read() happens
  printf("buffer size: %d\n", inputbuf->size);
  if (inputbuf->size == 0) {
    // no buffer to do, so register read event
    return;
  }
  // Now we get non-processed payload in BIO.
  dump_buffer(inputbuf);
  // we get buffer inside BIO so retry
  goto retry;
}
```

OpenSSL version is 1.1.1c.

Any comments and helps will be appreciated deeply.

Thank you.

Chul-Woong

**Comment by mattcaswell** [comment]
In the normal course of events this shouldn't happen. I'm struggling to think if a  scenario where it would. Are you using a custom BIO?

**Comment by cwyang** [comment]
Yes, I use libh2o (https://www.gihub.com/h2o/h2o) and it has following read_bio().
`sock->ssl->input.encrypted->bytes` BIO's input buffer.

```
static int read_bio(BIO *b, char *out, int len)
{
    h2o_socket_t *sock = BIO_get_data(b);

    if (len == 0)
        return 0;

    if (sock->ssl->input.encrypted->size == 0) {
        BIO_set_retry_read(b);
        return -1;
    }

    if (sock->ssl->input.encrypted->size < len) {
        len = (int)sock->ssl->input.encrypted->size;
    }
    memcpy(out, sock->ssl->input.encrypted->bytes, len);
    h2o_buffer_consume(&sock->ssl->input.encrypted, len);

    return len;
}
```

This symptom occurs at customer's production site so I cannot do more detailed analysis with info callback to find out whether:
* TLS alert is processed but does not consumed? --> improbable as processing means SSL3 records is fetched from BIO
* TLS alert is located at BIO but SSL internal status is changed to already error condition so `SSL_accept()` returns -1 without touching BIO's pending TLS alert. It's my guess and I barely understands openssl code.

**Comment by mattcaswell** [comment]
Without being able to reproduce this I think its going to be tough to track this one down. Ideally we would instrument the read_bio function you provided above to see if it is ever called in the scenario you are encountering - and if it is what does it return?

**Comment by cwyang** [comment]
Yes, it's not easy to resolve issue without reproduction.
I'll wait for other peoples comments while experimenting myself.
Thank you very much for your comments.

**Comment by cwyang** [comment]
Hi, I got some!

When SSL_accept() of openssl 1.1.1c returns (-1, WANT_READ) without processing pending TLS record with these results:
- SSL_get_state = 46 (TLS_ST_EARLY_DATA)
- SSL_in_init = 1
- SSL_pending = SSL_has_pending = 0
- SSL_get_shutdown = 2 (RECVD_SHUTDOWN)

It does not look like normal case with received shutdown at early data phase.
However, debug probes tells us above thing.
I'd be happy if this helps.

Take care,
Chul-Woong

**Comment by nhorman** [comment]
Marking as inactive to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
